src/units/abilities.cpp

```diff
@@ -1168,33 +1168,85 @@ void attack_type::modified_attacks(unsigned & min_attacks,
 	}
 }
 
-//Functions used for change damage_type list with damage
-static std::optional<std::string> select_damage_type(const unit_ability_list& abil_list, const std::string& type)
+//functions for select damage type who must be returned for each attribute used in damage_type()
+static std::optional<std::string> select_replacement_type(const unit_ability_list& abil_list)
 {
-	std::vector<std::string> type_list;
+	std::unordered_map<std::string, unsigned int> type_count;
+	unsigned int max = 0;
 	for(auto& i : abil_list) {
-		if(!(*i.ability_cfg)[type].str().empty()){
-			type_list.push_back((*i.ability_cfg)[type].str());
+		const config& c = *i.ability_cfg;
+		if(!c["replacement_type"].empty()) {
+			std::string type = c["replacement_type"].str();
+			unsigned int count = ++type_count[type];
+			if((count > max)) {
+				max = count;
+			}
+		}
+	}
+
+	if (type_count.empty()) return {};
+
+	std::vector<std::string> type_list;
+	for(auto& i : type_count){
+		if(i.second == max){
+			type_list.push_back(i.first);
 		}
 	}
-	if(type_list.size() >= 2){
+
+	if(type_list.size() > 1){
 		std::sort(type_list.begin(), type_list.end());
-		if(type_list.size() >= 3){
-			std::unordered_map<std::string, unsigned int> type_count;
-			for( const std::string& character : type_list ){
-				type_count[character]++;
+	}
+
+	if(!type_list.empty()){
+		return type_list.front();
+	}
+	return std::nullopt;
+}
+
+static std::optional<std::string> select_alternative_type(const unit_ability_list& abil_list, unit_ability_list resistance_abilities, const unit& u)
+{
+	std::map<std::string, std::optional<int>> type_res;
+	std::optional<int> max_res;
```

Comment:
replacedby std::string declaration except for std::optional<int> max_res; because i don't want see int initialised while none value is added.

Reply:
That's not the optional use I was talking about but I don't mind that.

There are two `std::optional<int>` uses in this line and the line above. They seem to complicate the code for no reason. There is an early return if `type_res` is empty after that `max_res` is set to a value and in the map it's set either way. I don't understand what you're worried about.

Reply:
For [f](std::map<std::string, std::optional<int>> type_res;) I agree and it was an error on my part but for max_res it is to avoid a display bug in an impossible situation let's take a dwarf guard with steadfeast, this guard is more sensitive to arcane than piercing , now what happens if steadfeas was modified to multiply the resistance by 20 instead of and remove max_value and filter_base_value, we then have a unit whose all resistances go beyond 100% (less than 0 in the code) , if I attack it then with an archer with a double pierce type, arcane; with my current coding the pre-combat window normally displays both types while if max_res is a normal int with an initial value of zero then arcane no longer appears since only resistance greater than 0 (less than or equal to 100 %) will be added, it may be stupid but it's to avoid issues where players will complain same if few probability

Reply:
Are you saying there is no minimum to initialize `max_res` with?

Reply:
yes, it is for no have minimum

Reply:
I mean you could just use INT_MIN but I suppose I don't mind optional then.

Reply:
INT_MINT?